README.md
isbn: ISBN Validation and Manipulation
======================================

[![Build Status](https://github.com/charukiewicz/hs-isbn/workflows/CI/badge.svg)](https://github.com/charukiewicz/hs-isbn/actions?query=workflow%3ACI)
[![Hackage](http://img.shields.io/hackage/v/isbn.svg)](https://hackage.haskell.org/package/isbn)
[![Hackage CI](https://matrix.hackage.haskell.org/api/v2/packages/isbn/badge)](https://matrix.hackage.haskell.org/package/isbn)
[![Stackage LTS](http://stackage.org/package/isbn/badge/lts)](http://stackage.org/lts/package/isbn)
[![Stackage Nightly](http://stackage.org/package/isbn/badge/nightly)](http://stackage.org/nightly/package/isbn)
[![Apache-2.0](https://img.shields.io/badge/license-Apache--2.0-blue)](https://www.apache.org/licenses/LICENSE-2.0.html)

All books published by major publishers since 1970 have an International Standard Book Number (ISBN) associated with them. Prior to 2007, all ISBNs issued were ten digit ISBN-10 format. Since 2007, new ISBNs have been issued in the thirteen digit ISBN-13 format. See the [ISBN Wikipedia article](https://en.wikipedia.org/wiki/International_Standard_Book_Number) for more information.

## Overview

This library provides data types and functions both validating and working with ISBNs. For general use, only importing the `Data.ISBN` module is required, as it reexports all functionality for validating and converting between ISBN-10 and ISBN-13 values. The specific implementations for validation are located in the `Data.ISBN.ISBN10` and `Data.ISBN.ISBN13` modules, respectively.

## Usage Example: Validating an ISBN and printing the error

```haskell
import Data.ISBN (ISBN, validateISBN, renderISBNValidationError)

validateUserSuppliedISBN :: Text -> Either Text ISBN
validateUserSuppliedISBN userIsbnInput =
    either (Left . renderISBNValidationError) Right (validateISBN userIsbnInput)


someValidISBN10 =
    validateUserSuppliedISBN "0345816021"    -- Right (ISBN10 "0345816021")

someValidISBN13 =
    validateUserSuppliedISBN "9780807014295" -- Right (ISBN13 "9780807014295")

tooShortISBN =
    validateUserSuppliedISBN "0-345-816"     -- Left "An ISBN must be 10 or 13 characters, not counting hyphens"

invalidISBN10 =
    validateUserSuppliedISBN "0-345-81602-3" -- Left "The supplied ISBN-10 is not valid"
```


## Development

This library is developed using a [Nix](https://nixos.org/nix/) shell. The environment is specified in `shell.nix`. You can enter the environment with `nix` installed via `nix-shell`. The `nix-shell` will install sandboxed copies of `cabal-install`, `ghcid`, `entr`, and `gnumake`, which are the utilities necessary to build the project, run the tests, and create a local copy of the documentation.

#### Entering the development environment and runnning tests

```
$ cd <path-to-repo>
$ nix-shell                       # assumes `nix` is installed
(nix-shell) $ make help           # show all of the make targets
(nix-shell) $ make tests-watch    # build the library and run the tests in `ghcid`
(nix-shell) $ make docs           # build a local copy of the haddock documentation
```


CHANGELOG.md
# Revision history for `isbn`

## 1.1.0.5 - Dec 26, 2023

- Bump version of `text` dependency to include versions up to 2.2 (exclusive)
- Add `flake.nix` file for Nix Flake-based development (via `nix develop`)
- Bump `nixpkgs.json` to `release-23.11` branch of Nixpkgs for `nix-shell` based development

## 1.1.0.4 - September 6, 2022

- Bumped version bound of `text` dependency to include `text-2.0.x`.

## 1.1.0.3 - January 10, 2022

- Changed imports of `Data.Text` to be qualified to prevent ambiguous import issues caused by changed to `Prelude` in GHC 9.x.

## 1.1.0.2 - November 13, 2020

- Built documentation using GHC 8.8.4
- Changed Nix shell development environment to use `fetchFromGitHub` instead of `fetchGit` for nixpkgs
- Changed `hackage-build` Makefile target to split `sdist` and `docs` outputs into separate subdirectories of `build/`
- Added `hackage-upload` and `hackage-upload-publish` Makefile targets to upload new package builds to Hackage

## 1.1.0.1 - July 30, 2020

- Changed implementation of `validateISBN` to make a copy of the input text, preventing space leaks
- Added some test cases for ISBN-13s with a '979' prefix

## 1.1.0.0 - July 28, 2020

- Changed exports in `Data.ISBN` module to not export the constructors of the `ISBN` data type
  - If constructors are needed, the `Data.Types.ISBN` module can be imported instead
- Added `isISBN10` and `isISBN13` functions to allow for checking of the opaque `ISBN` data type without the use of its value constructors

## 1.0.0.0 - May 17, 2020

- Initial release. Features:
  - General ISBN validation based off of text input, as well as specific ISBN-10 or ISBN-13 validation
  - Validation errors with human-friendly renderers, that can be used to display error messages to end users
  - ISBN-10 to ISBN-13 conversion and vise versa


test/Spec.hs
{-# LANGUAGE OverloadedStrings #-}

module Main (main) where

import           Data.ISBN
import           Data.ISBN.ISBN10
import           Data.Text        ( pack )

import           Data.Traversable

import           Test.Hspec


main :: IO ()
main = hspec $ do
    describe "Checking for valid ISBN 10 check digit characters" $ do
        let testValidChar char expecting =
              it ("can " ++ (if expecting then "accept" else "reject") ++ " '" ++ [char] ++ "'") $
                  isValidISBN10CheckDigit char `shouldBe` expecting

        testValidChar 'X' True
        testValidChar '1' True
        testValidChar '2' True
        testValidChar '3' True
        testValidChar '4' True
        testValidChar '5' True
        testValidChar '6' True
        testValidChar '7' True
        testValidChar '8' True
        testValidChar '9' True
        testValidChar '0' True
        testValidChar 'C' False
        testValidChar 'F' False
        testValidChar 'a' False
        testValidChar 'x' False
        testValidChar '-' False
        testValidChar '_' False
        testValidChar '*' False
        testValidChar '!' False
        testValidChar ' ' False

    describe "Checking ISBN 10 check digit confirmation" $ do
        let testCheckDigit isbn10string expecting =
              it ("can " ++ (if expecting then "confirm" else "reject ") ++ " '" ++ isbn10string ++ "'") $
                  confirmISBN10CheckDigit (pack isbn10string) `shouldBe` expecting

        testCheckDigit "0306406152" True
        testCheckDigit "030640615X" False
        testCheckDigit "0345816021" True
        testCheckDigit "080701429X" True
        testCheckDigit "0060899220" True
        testCheckDigit "2222222222" True
        testCheckDigit "4444444444" True
        testCheckDigit "9999999999" True
        testCheckDigit "999999999X" False

    describe "Validating possible ISBN-10s" $ do
        let testISBN10 isbn10 expecting =
                it ("can " ++ (case expecting of Right _ -> "validate"; Left _ -> "reject  ") ++ " '" ++ isbn10 ++ "'") $
                    validateISBN10 (pack isbn10) `shouldBe` expecting

        testISBN10 "0-345-81602-1" (Right $ unsafeToISBN10 "0345816021")
        testISBN10 "0345816021"    (Right $ unsafeToISBN10 "0345816021")
        testISBN10 "080701429X"    (Right $ unsafeToISBN10 "080701429X")
        testISBN10 "0-345-81602-0" (Left $ InvalidCheckDigit)
        testISBN10 "0-345-X1602-0" (Left $ IllegalCharactersInBody)
        testISBN10 "A-345-X1602-0" (Left $ IllegalCharactersInBody)
        testISBN10 "A2345-X1602-0" (Left $ InvalidInputLength)
        testISBN10 "345-81602-1"   (Left $ InvalidInputLength)


test/Data/ISBNSpec.hs
{-# LANGUAGE OverloadedStrings #-}

module Data.ISBNSpec (spec) where

import           Data.ISBN
import           Data.ISBN.ISBN10
import           Data.ISBN.ISBN13

import           Data.Text        ( pack )
import           Test.Hspec


spec :: Spec
spec = do
    describe "Testing ISBN validation" $ do
        let test_validateISBN isbn expecting =
                it (concat
                        [ "can "
                        , case expecting of Right _ -> "validate"; Left _ -> "reject  "
                        , " '"
                        , isbn
                        , "'"
                        , take (18 - length isbn) $ cycle " "
                        , "as '"
                        , show expecting
                        , "'"
                        ]) $
                    validateISBN (pack isbn) `shouldBe` expecting

        test_validateISBN "0345816021"        (Right (unsafeToISBN10 "0345816021"))
        test_validateISBN "0-807-01429-X"     (Right (unsafeToISBN10 "080701429X"))
        test_validateISBN "9780807014295"     (Right (unsafeToISBN13 "9780807014295"))
        test_validateISBN "978-0-306-40615-7" (Right (unsafeToISBN13 "9780306406157"))
        test_validateISBN "979-10-95546-00-9" (Right (unsafeToISBN13 "9791095546009"))
        test_validateISBN "0-345-816"         (Left InvalidISBNInputLength)
        test_validateISBN "X-345-81602-1"     (Left IllegalCharactersInISBN10Body)
        test_validateISBN "0-345-81602-B"     (Left IllegalCharacterAsISBN10CheckDigit)
        test_validateISBN "0-345-81602-3"     (Left InvalidISBN10CheckDigit)
        test_validateISBN "00000000000000"    (Left InvalidISBNInputLength)
        test_validateISBN "9780807014299"     (Left InvalidISBN13CheckDigit)
        test_validateISBN "0X00000000000"     (Left IllegalCharactersInISBN13Input)

    describe "Testing ISBN-10 to ISBN-13 conversion" $ do
        let test_convertISBN10toISBN13 isbn10 expecting =
                it ("can convert (" ++ show isbn10 ++ ") to (" ++ show expecting ++ ")") $
                    convertISBN10toISBN13 isbn10 `shouldBe` expecting

        test_convertISBN10toISBN13 (unsafeToISBN10 "0060899220") (unsafeToISBN13 "9780060899226")
        test_convertISBN10toISBN13 (unsafeToISBN10 "1519150245") (unsafeToISBN13 "9781519150240")

    describe "Testing ISBN-13 to ISBN-10 conversion" $ do
        let test_convertISBN13toISBN10 isbn13 expecting =
                it ("can convert (" ++ show isbn13 ++ ") to (" ++ show expecting ++ ")") $
                    convertISBN13toISBN10 isbn13 `shouldBe` expecting

        test_convertISBN13toISBN10 (unsafeToISBN13 "9780060899226") (Just $ unsafeToISBN10 "0060899220")
        test_convertISBN13toISBN10 (unsafeToISBN13 "9781519150240") (Just $ unsafeToISBN10 "1519150245")
        test_convertISBN13toISBN10 (unsafeToISBN13 "2222222222222") Nothing


test/Data/ISBN/ISBN10Spec.hs
{-# LANGUAGE OverloadedStrings #-}

module Data.ISBN.ISBN10Spec (spec) where

import           Data.ISBN.ISBN10

import           Data.Text        ( pack )
import           Test.Hspec


spec :: Spec
spec = do
    describe "Testing validity of ISBN-10 check digit characters" $ do
        let test_isValidISBN10CheckDigit char expecting =
              it ("can " ++ (if expecting then "accept" else "reject") ++ " '" ++ [char] ++ "'") $
                  isValidISBN10CheckDigit char `shouldBe` expecting

        test_isValidISBN10CheckDigit 'X' True
        test_isValidISBN10CheckDigit '1' True
        test_isValidISBN10CheckDigit '2' True
        test_isValidISBN10CheckDigit '3' True
        test_isValidISBN10CheckDigit '4' True
        test_isValidISBN10CheckDigit '5' True
        test_isValidISBN10CheckDigit '6' True
        test_isValidISBN10CheckDigit '7' True
        test_isValidISBN10CheckDigit '8' True
        test_isValidISBN10CheckDigit '9' True
        test_isValidISBN10CheckDigit '0' True
        test_isValidISBN10CheckDigit 'C' False
        test_isValidISBN10CheckDigit 'F' False
        test_isValidISBN10CheckDigit 'a' False
        test_isValidISBN10CheckDigit 'x' False
        test_isValidISBN10CheckDigit '-' False
        test_isValidISBN10CheckDigit '_' False
        test_isValidISBN10CheckDigit '*' False
        test_isValidISBN10CheckDigit '!' False
        test_isValidISBN10CheckDigit ' ' False

    describe "Testing ISBN-10 check digit confirmation" $ do
        let test_confirmISBN10CheckDigit isbn10string expecting =
              it ("can " ++ (if expecting then "confirm" else "reject ") ++ " '" ++ isbn10string ++ "'") $
                  confirmISBN10CheckDigit (pack isbn10string) `shouldBe` expecting

        test_confirmISBN10CheckDigit "0306406152" True
        test_confirmISBN10CheckDigit "030640615X" False
        test_confirmISBN10CheckDigit "0345816021" True
        test_confirmISBN10CheckDigit "080701429X" True
        test_confirmISBN10CheckDigit "0060899220" True
        test_confirmISBN10CheckDigit "2222222222" True
        test_confirmISBN10CheckDigit "4444444444" True
        test_confirmISBN10CheckDigit "9999999999" True
        test_confirmISBN10CheckDigit "999999999X" False

    describe "Testing ISBN-10 validation with errors" $ do
        let test_validateISBN10 isbn10 expecting =
                it (concat
                        [ "can "
                        , case expecting of Right _ -> "validate"; Left _ -> "reject  "
                        , " '"
                        , isbn10
                        , "'"
                        , take (14 - length isbn10) $ cycle " "
                        , "as '"
                        , show expecting
                        , "'"
                        ]) $
                    validateISBN10 (pack isbn10) `shouldBe` expecting

        test_validateISBN10 "0-345-81602-1" (Right $ unsafeToISBN10 "0345816021")
        test_validateISBN10 "0345816021"    (Right $ unsafeToISBN10 "0345816021")
        test_validateISBN10 "080701429X"    (Right $ unsafeToISBN10 "080701429X")
        test_validateISBN10 "0-345-81602-0" (Left ISBN10InvalidCheckDigit)
        test_validateISBN10 "0-345-X1602-0" (Left ISBN10IllegalCharactersInBody)
        test_validateISBN10 "A-345-X1602-0" (Left ISBN10IllegalCharactersInBody)
        test_validateISBN10 "A2345-X1602-0" (Left ISBN10InvalidInputLength)
        test_validateISBN10 "345-81602-1"   (Left ISBN10InvalidInputLength)


lib/Data/ISBN/Types.hs
module Data.ISBN.Types where

import           Data.Text


-- | Data type for representing ISBN values. Values of this type should be
-- created safely using 'Data.ISBN.validateISBN', which will produce 'ISBN10'
-- or 'ISBN13' values after validating the input.
--
-- The 'Data.ISBN.validateISBN10' and 'Data.ISBN.validateISBN13' functions can
-- also be used to only attempt to create ISBNs of a specific type.
--
-- When importing this data type, it is recommended not expose the
-- constructors, and instead use 'Data.ISBN.unsafeToISBN10' or
-- 'Data.ISBN.unsafeToISBN13' to coerce 'Data.Text.Text' values into 'ISBN'
-- values.
data ISBN
    = ISBN10 Text -- ^ An ISBN-10 value. Consists of 9 digits followed by a base-11 check digit (@0-9@ or @\'X\'@).
    | ISBN13 Text -- ^ An ISBN-13 value. Consists of 12 digits followed by a base-10 check digit (@0-9@).
    deriving (Show, Eq)



lib/Data/ISBN/ISBN10.hs
{-# LANGUAGE OverloadedStrings #-}

module Data.ISBN.ISBN10
    ( ISBN(..)
    , validateISBN10
      -- * Validation Errors
    , ISBN10ValidationError(..)
    , renderISBN10ValidationError
      -- * Helpers
    , confirmISBN10CheckDigit
    , calculateISBN10CheckDigitValue
    , isbn10CharToNumericValue
    , numericValueToISBN10Char
    , isValidISBN10CheckDigit
    , isNumericCharacter
    , isISBN10
      -- * Unsafe Coercion
    , unsafeToISBN10
    ) where

import           Control.Monad
import           Data.Char
import           Data.Text       as Text

import           Data.ISBN.Types ( ISBN (ISBN10) )



-- | Used to safely create 'ISBN10' values represented by the 'ISBN' data type.
-- Assumes that the 'Data.Text.Text' input is an ISBN-10 string, either with or
-- without hyphens.
--
-- Will return either a validated ISBN-10 or an 'ISBN10ValidationError', which can be
-- rendered as a descriptive string using 'renderISBN10ValidationError'.
--
-- /Examples:/
--
-- @
-- validateISBN10 "0-345-81602-1" == Right (ISBN10 "0345816021")
-- validateISBN10 "0345816021"    == Right (ISBN10 "0345816021")
-- validateISBN10 "0-807-01429-X" == Right (ISBN10 "080701429X")
-- validateISBN10 "0-345-816"     == Left ISBN10InvalidInputLength
-- validateISBN10 "X-345-81602-1" == Left ISBN10IllegalCharactersInBody
-- validateISBN10 "0-345-81602-B" == Left ISBN10IllegalCharacterAsCheckDigit
-- validateISBN10 "0-345-81602-3" == Left ISBN10InvalidCheckDigit
-- @
validateISBN10 :: Text -> Either ISBN10ValidationError ISBN
validateISBN10 input = do
    -- Make a copy of the text input before further manipulation to prevent
    -- space leaks if input text is a slice of a larger string
    let inputWithoutHyphens = Text.filter (/= '-') $ Text.copy input

    unless (Text.length inputWithoutHyphens == 10) $
        Left ISBN10InvalidInputLength

    let invalidBodyCharacters = Text.filter (not . isNumericCharacter) (Text.init inputWithoutHyphens)

    unless (Text.length invalidBodyCharacters == 0) $
        Left ISBN10IllegalCharactersInBody

    unless (isValidISBN10CheckDigit $ Text.last inputWithoutHyphens) $
        Left ISBN10IllegalCharacterAsCheckDigit

    unless (confirmISBN10CheckDigit inputWithoutHyphens) $
        Left ISBN10InvalidCheckDigit

    pure $ ISBN10 inputWithoutHyphens



-- | Possible validation errors resulting from ISBN-10 validation.
data ISBN10ValidationError
    = ISBN10InvalidInputLength           -- ^ The length of the input string is not 10 characters, not counting hyphens
    | ISBN10IllegalCharactersInBody      -- ^ The first nine characters of the ISBN-10 input contain non-numeric characters
    | ISBN10IllegalCharacterAsCheckDigit -- ^ The check digit of the ISBN-10 is not a valid character (@0-9@ or @\'X\'@)
    | ISBN10InvalidCheckDigit            -- ^ The check digit is not valid for the given ISBN-10
    deriving (Show, Eq)


-- | Convert an 'ISBN10ValidationError' into a human-friendly error message.
renderISBN10ValidationError :: ISBN10ValidationError -> Text
renderISBN10ValidationError validationError =
    case validationError of
        ISBN10InvalidInputLength ->
            "An ISBN-10 must be 10 characters, not counting hyphens"

        ISBN10IllegalCharactersInBody ->
            "The first nine characters of an ISBN-10 must all be numbers"

        ISBN10IllegalCharacterAsCheckDigit ->
            "The last character of the supplied ISBN-10 must be a number or the letter 'X'"

        ISBN10InvalidCheckDigit ->
            "The supplied ISBN-10 is not valid"


-- | Confirms that the check digit of an ISBN-10 is correct. Assumes that the
-- input consists of 9 numeric characters followed by a legal check digit
-- character (@0-9@ or @X@).
--
-- /Examples:/
--
-- @
-- confirmISBN10CheckDigit "0345816021" == True
-- confirmISBN10CheckDigit "080701429X" == True
-- @
confirmISBN10CheckDigit :: Text -> Bool
confirmISBN10CheckDigit isbn10 =
    calculateISBN10CheckDigitValue (Text.init isbn10) == isbn10CharToNumericValue (Text.last isbn10)


-- | Calculates an ISBN-10 check digit value using the standard check digit
-- calculation formula. Assumes that the input is 9 numeric characters. The
-- check digit value can be any number in the range 0 to 10, the last of
-- which is represented by the symbol \'X\' in an ISBN-10.
--
-- See: <https://en.wikipedia.org/wiki/International_Standard_Book_Number#ISBN-10_check_digits>
--
-- /Examples:/
--
-- @
-- calculateISBN10CheckDigitValue "034581602" == 1
-- calculateISBN10CheckDigitValue "080701429" == 10
-- @
calculateISBN10CheckDigitValue :: Text -> Int
calculateISBN10CheckDigitValue input =
    go 10 (unpack input) 0
      where
        go n charList acc =
            case charList of
              [] -> (11 - (acc `mod` 11)) `mod` 11
              c:clist -> go (n - 1) clist (acc + isbn10CharToNumericValue c * n)


-- | Converts an ISBN-10 character to a numeric value. Valid input characters
-- include @0-9@ as well as @X@.
isbn10CharToNumericValue :: Char -> Int
isbn10CharToNumericValue 'X' = 10
isbn10CharToNumericValue  c  = digitToInt c

-- | Converts a numeric value to an ISBN-10 character. Valid input values
-- are the numbers from 0 to 10.
numericValueToISBN10Char :: Int -> Char
numericValueToISBN10Char 10 = 'X'
numericValueToISBN10Char c  = Text.head $ pack $ show c


-- | Validates a character as a valid ISBN-10 check digit character.  ISBN-10
-- check digit characters include @0-9@ as well as the symbol @'X'@. The lowercase
-- letter \'x\' is not considered valid.
isValidISBN10CheckDigit :: Char -> Bool
isValidISBN10CheckDigit char = char `elem` ("1234567890X" :: String)

-- | Determines whether a character is numeric (e.g. in the range of @0-9@).
isNumericCharacter :: Char -> Bool
isNumericCharacter char = char `elem` ("1234567890" :: String)

-- | Determines whether an 'ISBN' value is an ISBN-10.
--
-- /Examples:/
--
-- @
-- isISBN10 (unsafeToISBN10 "0060899220")    == True
-- isISBN10 (unsafeToISBN13 "9780060899226") == False
-- @
--
-- /Since: 1.1.0.0/
isISBN10 :: ISBN -> Bool
isISBN10 (ISBN10 _) = True
isISBN10 _          = False


-- | Will create an 'ISBN10' value without any validation.
unsafeToISBN10 :: Text -> ISBN
unsafeToISBN10 = ISBN10


lib/Data/ISBN.hs
{-# LANGUAGE FlexibleContexts  #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE OverloadedStrings #-}


module Data.ISBN
    ( -- * Introduction
      -- $introduction

      -- * Documentation
      ISBN
      -- $isbn-import
    , renderISBN
      -- * ISBN Validation
    , validateISBN
      -- *** ISBN Validation Errors
    , ISBNValidationError(..)
    , renderISBNValidationError

      -- * Validating only ISBN-10
    , validateISBN10
      -- *** ISBN-13 Validation Errors
    , ISBN10ValidationError
    , renderISBN10ValidationError

      -- * Validating only ISBN-13
    , validateISBN13
      -- *** ISBN-10 Validation Errors
    , ISBN13ValidationError(..)
    , renderISBN13ValidationError

      -- * Conversion between ISBN-10 and ISBN-13
      -- $conversion
    , convertISBN10toISBN13
    , convertISBN13toISBN10
      -- * ISBN Helpers
    , isISBN10
    , isISBN13
      -- * Creating ISBN values without validation
      -- $unsafe
    , unsafeToISBN10
    , unsafeToISBN13
    ) where

import           Data.ISBN.ISBN10
import           Data.ISBN.ISBN13

import           Control.Monad
import           Data.Text        as Text

-- $introduction
--
-- This library contains tools for validating and working with
-- [International Standard Book Numbers (ISBNs)](https://en.wikipedia.org/wiki/International_Standard_Book_Number).
-- An ISBN value can be in either the format of a ten digit ISBN-10 or a
-- thirteen digit ISBN-13. This library uses a single 'ISBN' data type with two
-- value constructors to represent ISBN values, so that a single entry point can
-- be used to validate text inputs as either an 'ISBN10', 'ISBN13', or neither.
--
-- For most users, importing only the @Data.ISBN@ module is necessary, as this
-- module re-exports all of the data types and functions necessary for
-- validating, creating, and converting between ISBN values.


------------------------------------

-- $isbn-import
--
-- __NOTE:__ This module does not export the constructors in the @ISBN@ type.
-- For access to the constructors, import the @Data.Types.ISBN@ module.


------------------------------------


-- | Used to safely create 'ISBN' values. Assumes that the 'Data.Text.Text'
-- input is an ISBN-10 or ISBN-13 string, either with or without hyphens.
--
-- Will return either a validated @ISBN@ or an 'ISBNValidationError', which can be
-- rendered as a descriptive string using 'renderISBNValidationError'.
--
-- /Examples:/
--
-- @
-- validateISBN "0345816021"        == Right (ISBN10 "0345816021")
-- validateISBN "0-807-01429-X"     == Right (ISBN10 "080701429X")
-- validateISBN "9780807014295"     == Right (ISBN13 "9780807014295")
-- validateISBN "978-0-306-40615-7" == Right (ISBN13 "9780306406157")
-- validateISBN "0-345-816"         == Left InvalidISBNInputLength
-- validateISBN "X-345-81602-1"     == Left IllegalCharactersInISBN10Body
-- validateISBN "0-345-81602-B"     == Left IllegalCharacterAsISBN10CheckDigit
-- validateISBN "0-345-81602-3"     == Left InvalidISBN10CheckDigit
-- validateISBN "00000000000000"    == Left InvalidISBNInputLength
-- validateISBN "9780807014299"     == Left InvalidISBN13CheckDigit
-- validateISBN "0X00000000000"     == Left IllegalCharactersInISBN13Input
-- @
validateISBN :: Text -> Either ISBNValidationError ISBN
validateISBN isbn = do
    let isbn10result = validateISBN10 isbn
        isbn13result = validateISBN13 isbn

    case (isbn10result, isbn13result) of
        (Right isbn10, _) ->
            Right isbn10

        (_, Right isbn13) ->
            Right isbn13

        (Left ISBN10InvalidInputLength, Left ISBN13InvalidInputLength) ->
            Left InvalidISBNInputLength

        (Left ISBN10IllegalCharactersInBody, _) ->
            Left IllegalCharactersInISBN10Body

        (Left ISBN10IllegalCharacterAsCheckDigit, _) ->
            Left IllegalCharacterAsISBN10CheckDigit

        (_ , Left ISBN13IllegalCharactersInInput) ->
            Left IllegalCharactersInISBN13Input

        (Left ISBN10InvalidCheckDigit, _) ->
            Left InvalidISBN10CheckDigit

        (_, Left ISBN13InvalidCheckDigit) ->
            Left InvalidISBN13CheckDigit


-- | Convert an 'ISBN' value to a 'Text' string. Can be used when displaying an
-- ISBN in an application interface or before storing the plain ISBN text values
-- in a database.
--
-- 'ISBN' values created using 'validateISBN', 'validateISBN10', or
-- 'validateISBN13' will never contain hyphens.
--
-- /Examples:/
--
-- @
-- renderISBN (ISBN10 "080701429X")    == "080701429X"
-- renderISBN (ISBN13 "9780060899226") == "9780060899226"
-- @
renderISBN :: ISBN -> Text
renderISBN (ISBN10 i) = i
renderISBN (ISBN13 i) = i


-- | Possible validation errors resulting from ISBN validation. Can be
-- rendered as a descriptive error message using 'renderISBNValidationError'.
data ISBNValidationError
    = InvalidISBNInputLength             -- ^ The length of the input string is not 10 or 13 characters, not counting hyphens
    | IllegalCharactersInISBN10Body      -- ^ The first nine characters of the ISBN-10 input contain non-numeric characters
    | IllegalCharactersInISBN13Input     -- ^ The ISBN-13 input contains non-numeric characters
    | IllegalCharacterAsISBN10CheckDigit -- ^ The check digit of the ISBN-10 is not a valid character (@0-9@ or @\'X\'@)
    | InvalidISBN10CheckDigit            -- ^ The check digit is not valid for the given ISBN-10
    | InvalidISBN13CheckDigit            -- ^ The check digit is not valid for the given ISBN-13
    deriving (Show, Eq)



-- | Convert an 'ISBNValidationError' into a human-friendly error message.
renderISBNValidationError :: ISBNValidationError -> Text
renderISBNValidationError validationError =
    case validationError of
        InvalidISBNInputLength ->
            "An ISBN must be 10 or 13 characters, not counting hyphens"

        IllegalCharactersInISBN10Body ->
            "The first nine non-hyphen characters of an ISBN-10 must all be numbers"

        IllegalCharactersInISBN13Input ->
            "Every non-hyphen character of an ISBN-13 must be a number"

        IllegalCharacterAsISBN10CheckDigit ->
            "The last character of an ISBN-10 must be a number or the letter 'X'"

        InvalidISBN10CheckDigit ->
            "The supplied ISBN-10 is not valid"

        InvalidISBN13CheckDigit ->
            "The supplied ISBN-13 is not valid"




-- $conversion
--
-- ISBN values can be converted from ISBN-10 to ISBN-13 and vise versa.


-- | Convert an ISBN-10 to an ISBN-13. Since all ISBN-10s can be converted to
-- ISBN-13s, this operation cannot fail.
--
-- /Example:/
--
-- @
-- convertISBN10toISBN13 (unsafeToISBN10 "0060899220") == ISBN13 "9780060899226"
-- @
convertISBN10toISBN13 :: ISBN -> ISBN
convertISBN10toISBN13 isbn10 =
    unsafeToISBN13 $ isbn13Body <> isbn13CheckDigit
      where
        isbn13CheckDigit = Text.singleton . numericValueToISBN13Char $ calculateISBN13CheckDigitValue isbn13Body
        isbn13Body = "978" <> isbn10Body
        isbn10Body = Text.init $ renderISBN isbn10


-- | Convert an ISBN-13 to an ISBN-10. Since only ISBN-13s starting with '978'
-- can be converted, this operation may fail.
--
-- /Example:/
--
-- @
-- convertISBN13toISBN10 (unsafeToISBN13 "9780060899226") == Just (ISBN10 "0060899220")
-- @
convertISBN13toISBN10 :: ISBN -> Maybe ISBN
convertISBN13toISBN10 isbn13 = do
    let isbn13Text = renderISBN isbn13
    unless ("978" `isPrefixOf` isbn13Text)
        Nothing -- "Only ISBN-13s that begin with '978' can be converted to ISBN-10s"

    let isbn10Body = Text.init $ Text.drop 3 isbn13Text
        isbn10CheckDigit = Text.singleton . numericValueToISBN10Char $ calculateISBN10CheckDigitValue isbn10Body

    pure $ unsafeToISBN10 $ isbn10Body <> isbn10CheckDigit

-- $unsafe
--
-- In most cases, creating 'ISBN10' and 'ISBN13' values should be performed
-- using the 'validateISBN', 'validateISBN10', or 'validateISBN13' functions,
-- which ensure the ISBN values they produce are valid.
--
-- The functions below allow for the creation of ISBN values without any
-- validation. They should only be used in specific cases. For example, when
-- loading already-validated ISBN values stored in a text column in a database.
